tests/unit/test_task_queue.py

```diff
@@ -0,0 +1,172 @@
+# THIS FILE IS PART OF THE CYLC SUITE ENGINE.
+# Copyright (C) NIWA & British Crown (Met Office) & Contributors.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+# Tests for the task queue module
+
+import pytest
+from unittest.mock import Mock
+from collections import Counter
+from cylc.flow.task_queue import TaskQueue, Limiter
+from cylc.flow.task_state import TASK_STATUS_PREPARING
+
+
+MEMBERS = {"a", "b", "c", "d", "e", "f"}
+ACTIVE = Counter(["a", "a", "d"])
+
+
+@pytest.mark.parametrize(
+    "limit,task_name,expected",
+    [
+        (  # limited: limit reached
+            3,
+            "c",
+            True
+        ),
+        (  # not limited: not a member
+            3,
+            "g",
+            False
+        ),
+        (  # not limited: limit not reached
+            4,
+            "c",
+            False
+        )
+    ]
+)
+def test_limiter(limit, task_name, expected):
+    """Test the queue limiter."""
+    itask = Mock()
+    itask.tdef.name = task_name
+    limiter = Limiter(limit, MEMBERS)
+    assert limiter.is_limited(itask, ACTIVE) == expected
+
+
+def test_limiter_adopt():
+    """Test queue limiter adoption."""
+    limiter = Limiter(3, MEMBERS)
+    itask = Mock()
+    itask.tdef.name = "g"
+    # not limited: not a member
+    assert not limiter.is_limited(itask, ACTIVE)
+    limiter.adopt(["g"])
+    # limited: is now a member
+    assert limiter.is_limited(itask, ACTIVE)
+
+
+ALL_TASK_NAMES = [
+    "o1", "o2", "o3", "o4", "o5", "o6", "o7",
+    "s1", "s2", "s3", "s4", "s5",
+    "b1", "b2", "b3", "b4", "b5",
+    "foo"
+]
+
+
+DESCENDANTS = {
+    "root": ALL_TASK_NAMES + ["BIG", "SML", "OTH", "foo"],
+    "BIG": ["b1", "b2", "b3", "b4", "b5"],
+    "SML": ["s1", "s2", "s3", "s4", "s5"],
+    "OTH": ["o1", "o2", "o3", "o4", "o5", "o6", "o7"]
+}
+
+
+QCONFIG = {
+    "type": None,
+    "default": {
+        "limit": 6,
+        "members": []  # (populated with all tasks, in TaskQueue init)
+    },
+    "big": {
+        "members": ["BIG", "foo"],
+        "limit": 2
+    },
+    "sml": {
+        "members": ["SML", "foo"],
+        "limit": 3
+    }
+}
+
+
+READY_TASK_NAMES = ["b3", "s4", "o2", "s3", "b4", "o3", "o4", "o5", "o6", "o7"]
+
+
+@pytest.mark.parametrize(
+    "queue_type,"
+    "active,"
+    "expected_released,"
+    "expected_still_queued,"
```

Comment:
Are these commas meant to be inside the quotes? Doesn't seem to have broken the test?

Reply:
Yes (weirdly). It's the `@pytest.mark.parametrize()` arg names, which is a comma-separated string, but split over multiple lines here.